### PR TITLE
[translator/jaeger] Parse and set Jaeger status as the OpenTelemetry specification defines

### LINKED ISSUE #6675: Jaeger translation of span status to pdata incorrectly parsed as integer
## Description

The `jaegerthrift` receiver parses the `otel.status_code` tag as an integer.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/3de607f6cd38673924397f8a3c9158d89b491207/pkg/translator/jaeger/jaegerproto_to_traces.go#L261

The [OpenTelemetry specification](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/trace/sdk_exporters/non-otlp.md#span-status) specifies this tag to be a string of `OK` or `ERROR` values.

Currently, this is correctly marking spans as errors because the jaeger span also has the `error` tag set to indicate the error and a span an appropriately tagged `otel.span_status` field will fail to parse and not override this status code. However, if, [like in the Go implementation](https://github.com/open-telemetry/opentelemetry-go/issues/2439), the tag is infact an integer and the integer does not match the `pdata` definitions of status, it is possible for an error spans to be marked successful.

## Proposal

- Parse `OK` and `ERROR` strings and translate to appropriate `pdata` status.
- Fallback to parsing `otel.status_code` only when the `error` tag is not set. Otherwise there is a chance the first parse of `error` is overwritten.

### PR DESCRIPTION
**Pull request by MrAlias**
The OpenTelemetry specification defines [the OpenTelemetry span status communicated with Jaeger data types as an `otel.status_code` tag having a string value of `ERROR` or `OK`](https://github.com/open-telemetry/opentelemetry-specification/blob/v1.8.0/specification/trace/sdk_exporters/non-otlp.md#span-status). Currently, this is incorrectly parsed and transmitted as an integer representation of the pdata span status value. This changes that to comply with the OpenTelemetry specification.

Additionally, the error tag for Jaeger spans takes precedence over the OpenTelemetry status value in defining if a span is errored. This changes the Jaeger proto to traces transformation to not overwrite the status with the OpenTelemetry status.

Resolves #6675

### CODE REVIEW

File: pkg/translator/jaeger/constants.go [hunk 1]
```diff
@@ -18,7 +18,24 @@ import (
 	"errors"
 )
 
+const (
+	// Integer types use two's complement arithmetic.
+	maxUint = ^uint(0)
+	maxInt  = int64(maxUint >> 1)
+	minInt  = -maxInt - 1
```
Comment (bogdandrutu) [inline pkg/translator/jaeger/constants.go:25]: Why not using "math" package? https://pkg.go.dev/math#pkg-constants `math.MaxInt`?

File: pkg/translator/jaeger/jaegerproto_to_traces.go [hunk 2]
```diff
@@ -252,30 +270,55 @@ func setInternalSpanStatus(attrs pdata.AttributeMap, dest pdata.SpanStatus) {
...
+	return "", false
+}
+
+// Allow tests to check int range errors.
+var validIntRange = func(val int64) error {
+	if val > maxInt || val < minInt {
+		return fmt.Errorf("%w: %d", errRange, val)
+	}
+	return nil
+}
```
Comment (bogdandrutu) [inline pkg/translator/jaeger/jaegerproto_to_traces.go:292]: Where does this requirement with "int" comes from? It is extremely bad design to validate based on the arch type (32/64). It means that if you are running a 32bit version of the collector for the same received data you may get errors that you don't get on a 64bit version of the collector.

**Review by jpkrohling — COMMENTED** [review]
I'm also confused by the math, once that's clarified, I'll run a manual e2e test.

File: pkg/translator/jaeger/jaegerproto_to_traces.go (hunk 2 above)
Comment (MrAlias) [reply pkg/translator/jaeger/jaegerproto_to_traces.go:292]: Sure, I'll revert to just validating int32. :+1:

File: pkg/translator/jaeger/jaegerproto_to_traces.go [hunk 3]
```diff
@@ -252,30 +271,47 @@ func setInternalSpanStatus(attrs pdata.AttributeMap, dest pdata.SpanStatus) {
...
+// the attrVal or the value is outside the bounds of an int representation.
+func codeFromAttr(attrVal pdata.AttributeValue) (int, error) {
+	var val int
 	switch attrVal.Type() {
 	case pdata.AttributeValueTypeInt:
-		codeVal = attrVal.IntVal()
+		i := attrVal.IntVal()
+		if i > math.MaxInt32 || i < math.MinInt32 {
+			return 0, fmt.Errorf("%w: %d", errRange, val)
+		}
```
Comment (bogdandrutu) [inline pkg/translator/jaeger/jaegerproto_to_traces.go:297]: Do we really care about this check? What value does it bring? Does all this complication comes from the fact that `tracetranslator.StatusCodeFromHTTP` accepts int?

Comment (MrAlias) [reply pkg/translator/jaeger/jaegerproto_to_traces.go:297]: AFAIK, yes. It was there before and I didn't want to introduce an unreported truncation error by just casting to an `int`.

Comment (MrAlias) [reply pkg/translator/jaeger/jaegerproto_to_traces.go:297]: From async conversation: let's update `tracetranslator.StatusCodeFromHTTP` to accept an `int64` in a separate PR an avoid this entirely.

Comment (MrAlias) [reply pkg/translator/jaeger/jaegerproto_to_traces.go:297]: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6699

File: CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,8 @@
 
 ## 🧰 Bug fixes 🧰
 
+- `jaeger` receiver/exporter: Parse/set Jaeger status as OTel spec defines (#6682)
```
Comment (dmitryax) [inline CHANGELOG.md:11]: I agree it's more like a bug fix, but can be a breaking change for those who rely on the OpenCensus style status codes. Should we move it to the _Breaking changes_ section?

**Review by dmitryax — APPROVED** [review]

File: CHANGELOG.md (hunk 4 above)
Comment (MrAlias) [reply CHANGELOG.md:11]: sure thing

**Comment by bogdandrutu** [comment]
Please rebase

**Comment by MrAlias** [comment]
> Please rebase

@bogdandrutu PTAL

**Review by jpkrohling — COMMENTED** [review]
I did a manual test and it looks good. The test included the Jaeger receiver, the Jaeger exporter, and the logging exporter, and I verified that applications using the Jaeger convention (such as the Hot ROD demo) got converted to the internal value and back, both by checking the output of the logging exporter (jaeger -> internal -> OTLP) and Jaeger UI (jaeger -> internal -> jaeger).

```
Span #19
    Trace ID       : 00000000000000005f11a21978859641
    Parent ID      : 61618f8752981d93
    ID             : 398d42739cbddbb3
    Name           : GetDriver
    Kind           : SPAN_KIND_CLIENT
    Start time     : 2021-12-14 11:01:05.999406 +0000 UTC
    End time       : 2021-12-14 11:01:06.026823 +0000 UTC
    Status code    : STATUS_CODE_ERROR
    Status message : 
```

image

That all said, the only thing preventing me from approving is the empty file under "results". What's its role? Should it be there?

**Comment by jpkrohling** [comment]
Given that @bogdandrutu has left a few comments as well, I'll wait for his feedback until tomorrow before merging.

**Pull request merged**